File: tree/leetcode144_preorder/iteration.py
# 144. Binary Tree Preorder Traversal
# https://leetcode.com/problems/binary-tree-preorder-traversal/description/
# Given the root of a binary tree, return the preorder traversal of its nodes' values.

# Preorder: root -> left subtree -> right subtree


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution1(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        result = []
        stack = []
        current = root

        while stack or current:
            while current:
                result.append(current.val)
                stack.append(current)
                current = current.left

            current = stack.pop()
            current = current.right

        return result

File: tree/leetcode144_preorder/iteration.js
// 144. Binary Tree Preorder Traversal
// https://leetcode.com/problems/binary-tree-preorder-traversal/description/
// Given the root of a binary tree, return the preorder traversal of its nodes' values.

// Preorder: root -> left subtree -> right subtree

// Iterative
// Time complexity: O(n), because go through each node
// Space complexity: O(h), h is the height of the tree // !
var preorderTraversal = function (root) {
  let result = []; // save the result
  let stack = []; //  temporary store the current node
  let current = root;

  // if current is not null or there's still elements in the stack array
  while (current || stack.length > 0) {
    while (current) {
      result.push(current.val); // add the value of current node to the result
      stack.push(current); // store current node
      current = current.left; // move to the left subtree
    }
    current = stack.pop(); // once current is null, get the nearest node
    current = current.right; // move to the right subtree
  }
  return result;
};